webapp/packages/plugin-data-spreadsheet-new/src/DataGrid/DataGridSelection/useGridSelectionContext.tsx

```diff
@@ -35,72 +35,114 @@ export function useGridSelectionContext(modelResultData: IDatabaseDataResult | n
 
+  const isRowSelected = useCallback((rowIdx: number) => {
+    if (!modelResultData) {
+      throw new Error('Model result data must be provided');
+    }
+
+    const columnsLength = (modelResultData.data as SqlResultSet).columns?.length || 0;
+
+    return selectedCells.get(rowIdx)?.length === columnsLength;
+  }, [modelResultData, selectedCells]);
+
+  const unSelectRow = useCallback((rowIdx: number) => {
+    selectedCells.delete(rowIdx);
+  }, [selectedCells]);
+
   const selectRange = useCallback(action((idx: number, rowIdx: number) => {
-    if (!currentSelectedCell) {
+    if (!lastSelectedCell) {
       throw new Error('Current selected cell must be provided');
     }
 
-    const rowRange = rowIdx - currentSelectedCell.rowIdx;
-    const columnRange = idx - currentSelectedCell.idx;
+    selectedCells.clear();
 
-    // TODO we don't need this, will refactor
-    const isColumnToBottom = columnRange > 0;
-    const isRowToBottom = rowRange > 0;
+    const left = Math.min(idx, lastSelectedCell.idx);
+    const right = Math.max(idx, lastSelectedCell.idx);
+    const top = Math.min(rowIdx, lastSelectedCell.rowIdx);
+    const bottom = Math.max(rowIdx, lastSelectedCell.rowIdx);
 
-    selectedCells.clear();
+    const rowSelection = [];
+    for (let colIdx = left; colIdx <= right; colIdx++) {
+      rowSelection.push(colIdx);
+    }
+    for (let rowIdx = top; rowIdx <= bottom; rowIdx++) {
+      selectedCells.set(rowIdx, rowSelection);
+    }
+  }), [selectedCells, lastSelectedCell]);
 
-    for (let column = 0; column <= Math.abs(columnRange); column++) {
-      const idx = isColumnToBottom ? column + currentSelectedCell.idx : currentSelectedCell.idx - column;
-      if (idx === 0) {
-        continue;
-      }
-      for (let row = 0; row <= Math.abs(rowRange); row++) {
-        const rowIdx = isRowToBottom ? row + currentSelectedCell.rowIdx : currentSelectedCell.rowIdx - row;
+  const isSelected = useCallback((idx: number, rowIdx: number) => {
+    if (!selectedCells.has(rowIdx)) {
+      return false;
+    }
 
-        const hash = getHash(idx, rowIdx);
-        selectedCells.set(hash, { idx, rowIdx });
-      }
+    const rowSelection = selectedCells.get(rowIdx)!;
+    return rowSelection.indexOf(idx) !== -1;
+  },
+  [selectedCells]);
+
+  const unSelect = useCallback((idx: number, rowIdx: number) => {
+    if (!selectedCells.has(rowIdx)) {
+      return false;
+    }
+
+    const rowSelection = selectedCells.get(rowIdx)!;
+    const targetIndex = rowSelection.indexOf(idx);
+
+    if (targetIndex !== -1) {
+      rowSelection.splice(targetIndex, 1);
+      selectedCells.set(rowIdx, rowSelection);
+      return true;
     }
-  }), [selectedCells, currentSelectedCell]);
+    return false;
+  }, [selectedCells]);
+
+  const selectCell = useCallback((idx: number, rowIdx: number) => {
+    const rowSelection = selectedCells.get(rowIdx);
+    if (rowSelection === undefined) {
+      return;
+    }
```

Comment:
why cell wasn't selected after function execution? no error message, no description
correct logic is to create an array, place in it `idx` and set it to `selectedCells` in case when row does not exist in `selectedCells`